Simplenote/Controllers/PublishController.swift

```diff
@@ -0,0 +1,112 @@
+import Foundation
+
+@objc
+class PublishController: NSObject {
+    private var callbackMap = [String: PublishListenWrapper]()
+    private let timerFactory: TimerFactory
+    private let publishListenerFactory: PublishListenerFactory
+    private var timer: Timer?
+
+    init(timerFactory: TimerFactory = TimerFactory(), publishListenerFactory: PublishListenerFactory = PublishListenerFactory()) {
+        self.timerFactory = timerFactory
+        self.publishListenerFactory = publishListenerFactory
+    }
+
+    func updatePublishState(for note: Note, to published: Bool, completion: @escaping (PublishState) -> Void) {
+        if note.published == published {
+            return
+        }
+
+        callbackMap[note.simperiumKey] = publishListenerFactory.publishListenerWrapper(note: note, block: completion, expiration: Date())
+
+        timer = timerFactory.scheduledTimer(with: Constants.timeOut, completion: {
+            self.removeExpiredCallbacks()
+        })
+
+        changePublishState(for: note, to: published)
+
+        published ?
+            callbackMap[note.simperiumKey]?.update(to: .publishing) : callbackMap[note.simperiumKey]?.update(to: .unpublishing)
+    }
+
+    @objc(didReceiveUpdateFromSimperiumForKey:)
+    func didReceiveUpdateFromSimperium(for key: String) {
+        guard var wrapper = callbackMap[key] else {
+            return
+        }
+
+        wrapper.handleListenResponse()
+    }
+
+    private func changePublishState(for note: Note, to published: Bool) {
+        note.published = published
+        note.modificationDate = Date()
+        SPAppDelegate.shared().save()
+    }
+
+    private func removeListenerCallback(for key: String) {
+        callbackMap.removeValue(forKey: key)
+    }
+
+    private func removeExpiredCallbacks() {
+        for wrapper in callbackMap {
+            if wrapper.value.isExpired {
+                callbackMap.removeValue(forKey: wrapper.key)
+            }
+        }
+
+        if !callbackMap.isEmpty {
+            timer = timerFactory.scheduledTimer(with: Constants.timeOut, completion: {
+                self.removeExpiredCallbacks()
+            })
+        }
+    }
+}
+
+enum PublishState {
+    case publishing
+    case published
+    case unpublishing
+    case unpublished
+}
+
+struct PublishListenWrapper {
+    let note: Note
+    let block: (PublishState) -> Void
+    var expiration: Date
```

Comment:
This one should probably be immutable, and set just once (after creation).

The goal of this structure is to listen for either **Success** or **Failure** toggling the Pubilshing state (rather than subscribing to all of the PubilshState changes)